File: buerguer-pig/HamburgesasEIngredientes-Java-IvanRC/src/main/java/validator/Validator.java
package validator;

import exception.ingredienteException.IngredienteException;
import models.Hamburgesa;
import models.Ingrediente;

import java.util.ArrayList;
import java.util.List;

public class Validator {

    public static boolean validateNumber(String number, Exception exception, Number min) throws Exception {
        if(min instanceof Integer){
            var regex = "-?[0-9]*";
            if(!number.matches(regex)){
                throw exception;
            }
            if(Integer.parseInt(number) < Integer.parseInt(min.toString())){
                throw exception;
            }
        }else{
            if(min instanceof Double){
                var regex = "-?[0-9]*.[0-9]*";
                if(!number.matches(regex)){
                    throw exception;
                }
                if(Double.parseDouble(number) < Double.parseDouble(min.toString())){
                    throw exception;
                }
            }
        }
        return true;
    }

    public static boolean validateString(String text, Exception exception) throws Exception {
        if(text.isEmpty()){
            throw exception;
        }
        return true;
    }

    public static boolean validateIngrediente(Ingrediente ingrediente) throws Exception {
        validateNumber(String.valueOf(ingrediente.getId()), new IngredienteException.IngredienteBadRequestException(String.valueOf(ingrediente.getId())), 1);
        validateString(ingrediente.getNombre(), new IngredienteException.IngredienteBadRequestException(ingrediente.getNombre()));
        validateNumber(String.valueOf(ingrediente.getPrecio()), new IngredienteException.IngredienteBadRequestException(String.valueOf(ingrediente.getPrecio())), 0.0);
        return true;
    }

    public static boolean validateHamburgesa(Hamburgesa hamburgesa) throws Exception {
        validateString(hamburgesa.getNombre(), new IngredienteException.IngredienteBadRequestException(hamburgesa.getNombre()));
        List<Ingrediente> ingredientes = hamburgesa.getIngredientes();
        for(int i = 0; i<ingredientes.size(); i++){
            validateIngrediente(ingredientes.get(i));
        }
        return true;
    }

}

File: buerguer-pig/HamburgesasEIngredientes-Kotlin-IvanRC/src/main/kotlin/validator/Validator.kt
package validator

import exception.hamburgesaException.HamburgesaBadRequestException
import exception.ingredienteException.IngredienteBadRequestException
import model.Hamburgesa
import model.Ingrediente
import repository.Ingrediente.IngredienteRepository

fun String.validate(exception: Exception): Boolean{
    require(this.isNotEmpty()){
        throw exception
    }
    return true
}

fun String.validateNumber(exception: Exception, min: Number): Boolean{
    if(min is Int) {
        val regex = Regex("-?[0-9]*")
        require(this.matches(regex)){
            throw exception
        }
        require(this.toInt() >= min) {
            throw exception
        }
    }else{
        if(min is Double) {
            val regex = Regex("-?[0-9]*.[0-9]*")
            require(this.matches(regex)){
                throw exception
            }
            require(this.toDouble() >= min) {
                throw exception
            }
        }
    }
    return true
}

fun Ingrediente.validate(): Boolean{
    this.id.toString().validateNumber(IngredienteBadRequestException("id = "+this.id.toString()), 1)
    this.nombre.validate(IngredienteBadRequestException("nombre = ${this.nombre}"))
    this.precio.toString().validateNumber(IngredienteBadRequestException("precio = ${this.precio}"), 0.0)
    return true
}

fun Hamburgesa.validate(): Boolean{
    this.nombre.validate(HamburgesaBadRequestException("nombre = ${this.nombre}"))
    this.ingredientes.forEach {
        it.validate()
    }
    return true
}

fun validarListaDeIngredientes(ingredientes: List<Int>, ingredienteRepo: IngredienteRepository): Boolean {
    ingredientes.forEach{
        val ingrediente = ingredienteRepo.findById(it)
        require(ingrediente != null){
            throw HamburgesaBadRequestException("ingrediente de id = $it")
        }
    }
    return true
}

fun String.validateFecha(exception: Exception): Boolean{
    val regex = Regex("[0-9]{4}-[0-9]{2}-[0-9]{2}")
    require(this.matches(regex)){
        throw exception
    }
    return true
}